pycontrails/core/models.py

```diff
@@ -731,25 +762,271 @@ def interpolate_met(
-        raise ValueError(f"No variable key {vector_key} in ``vector`` and ``met`` is None")
+        raise ValueError(f"No variable key '{vector_key}' in 'vector' and 'met' is None")
+
+    if met_key in ("q", "specific_humidity"):
+        mda, log_applied = _extract_q(met, met_key, q_method)
+        out = interpolate_gridded_specific_humidity(
+            mda, vector, q_method, log_applied, **interp_kwargs
+        )
+
+    else:
+        try:
+            mda = met[met_key]
+        except KeyError as exc:
+            raise KeyError(f"No variable key '{met_key}' in 'met'.") from exc
+
+        out = vector.intersect_met(mda, **interp_kwargs)
+
+    vector[vector_key] = out
+    return out
+
+
+def _extract_q(met: MetDataset, met_key: str, q_method: str) -> tuple[MetDataArray, bool]:
+    """Extract specific humidity from ``met`` :class:`MetDataset`.
+
+    Parameters
+    ----------
+    met : MetDataset
+        Met data
+    met_key : str
+        Key of specific humidity in ``met``. Typically either ``"q"`` or ``"specific_humidity"``.
+    q_method : str
+        Method to use for interpolating specific humidity.
+
+    Returns
+    -------
+    mda : MetDataArray
+        Specific humidity data
+    log_applied : bool
+        Whether a log transform was applied to ``mda``.
+    """
+    if q_method != "log-q-log-p":
+        try:
+            return met[met_key], False
+        except KeyError as exc:
+            raise KeyError(f"No variable key '{met_key}' in 'met'.") from exc
+
+    try:
+        return met["log_specific_humidity"], True
+    except KeyError:
+        warnings.warn(
+            "No variable key 'log_specific_humidity' in 'met'. "
+            "Falling back to 'specific_humidity'. "
+            "Computation will be faster if 'log_specific_humidity' is provided."
+        )
 
     try:
-        mda = met[met_key]
+        return met[met_key], False
     except KeyError as exc:
-        raise KeyError(f"No variable key {met_key} in ``met``") from exc
+        raise KeyError(f"No variable key '{met_key}' in 'met'.") from exc
+
+
+def _prepare_q(  # type: ignore[return-value]
+    mda: MetDataArray, level: npt.NDArray[np.float_], q_method: str, log_applied: bool
+) -> tuple[MetDataArray, npt.NDArray[np.float_]]:
+    """Prepare specific humidity for interpolation with experimental ``q_method``.
+
+    Parameters
+    ----------
+    mda : MetDataArray
+        MetDataArray of specific humidity.
+    level : npt.NDArray[np.float_]
+        Levels to interpolate to, [:math:`hPa`].
+    q_method : str
+        One of ``"log-q-log-p"`` or ``"cubic-spline"``.
+    log_applied : bool
+        Whether a log transform was applied to ``mda``.
+
+    Returns
+    -------
+    mda : MetDataArray
+        MetDataArray of specific humidity transformed for interpolation.
+    level : npt.NDArray[np.float_]
+        Transformed levels for interpolation.
+    """
+    da = mda.data
+
+    if q_method == "log-q-log-p":
```

Comment:
it feels like there should be a separate function for each of these, although I guess that starts to get verbose.

Reply:
bc6ff61fc94886d2ec84c6af329e1ebd21bcb31d